osbuild/util/containers.py

```diff
@@ -110,21 +116,42 @@ def container_source(image):
     container_format = image["data"]["format"]
     image_name = image["data"]["name"]
 
-    if container_format not in ("dir", "oci-archive"):
+    if container_format not in ("dir", "oci-archive", "containers-storage"):
         raise RuntimeError(f"Unknown container format {container_format}")
 
-    with tempfile.TemporaryDirectory() as tmpdir:
-        tmp_source = os.path.join(tmpdir, "image")
-
-        if container_format == "dir" and image["manifest-list"]:
-            # copy the source container to the tmp source so we can merge the manifest into it
-            subprocess.run(["cp", "-a", "--reflink=auto", image_filepath, tmp_source], check=True)
-            merge_manifest(image["manifest-list"], tmp_source)
-        else:
-            # We can't have special characters like ":" in the source names because containers/image
-            # treats them special, like e.g. /some/path:tag, so we make a symlink to the real name
-            # and pass the symlink name to skopeo to make it work with anything
-            os.symlink(image_filepath, tmp_source)
-
-        image_source = f"{container_format}:{tmp_source}"
-        yield image_name, image_source
+    if container_format == "containers-storage":
+        storage_conf = image["data"]["storage"]  # should be set
+        driver = storage_conf.get("driver", "overlay")
+
+        # use `/containers/storage` for the containers-storage bind mount
+        storage_path = os.path.join(os.sep, "containers", "storage")
+        os.makedirs(storage_path, exist_ok=True)
+
+        with MountGuard() as mg:
+            # NOTE: the ostree.deploy.container needs explicit `rw` access to
+            # the containers-storage store even when bind mounted.
+            mg.mount(image_filepath, storage_path, rw=True)
+
+            image_source = f"{container_format}:[{driver}@{storage_path}+/run/containers/storage]{image_name}"
+            yield image_name, image_source
+
+            if driver == "overlay":
+                # NOTE: the overlay sub-directory isn't always released,
+                # so we need to force unmount it
+                subprocess.run(["umount", "-f", os.path.join(storage_path, "overlay")], check=False)
```

Comment:
I wasn't sure at first if this was better than making a new type of input but looking at it now I'm in favour of keeping it as a container input.  It makes sense because stages will be able to use this the same way regardless of format/transport.  So a stage gets an input and can, for example, run `skopeo copy dir:container.tar ...` or `skopeo copy containers-storage:[driver@path...]image ...` and it would _just work_.

The `org.osbuild.container-deploy` stage will have to have a separate (simpler) behaviour for this case because it wont need to copy the container into the storage.

Reply:
One thing though, let's separate these two conditional blocks into functions for readability (and unit testing).

Reply:
Separate functions makes sense